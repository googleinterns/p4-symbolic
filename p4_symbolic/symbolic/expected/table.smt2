; 
(set-info :status unknown)
(declare-fun standard_metadata.ingress_port () (_ BitVec 9))
(assert
 (let (($x46 (= standard_metadata.ingress_port (_ bv1 9))))
 (or (or false (= standard_metadata.ingress_port (_ bv0 9))) $x46)))
(assert
 (let (($x71 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv1 1))))))
 (let (($x78 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv0 1))))))
 (let ((?x81 (ite $x78 0 (ite $x71 1 (- 1)))))
 (and (and (not (or false false)) (or (or true $x71) $x78)) (= ?x81 (- 1)))))))
(check-sat)

; 
(set-info :status unknown)
(declare-fun standard_metadata.ingress_port () (_ BitVec 9))
(assert
 (let (($x46 (= standard_metadata.ingress_port (_ bv1 9))))
 (or (or false (= standard_metadata.ingress_port (_ bv0 9))) $x46)))
(assert
 (let (($x71 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv1 1))))))
 (let (($x78 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv0 1))))))
 (let ((?x81 (ite $x78 0 (ite $x71 1 (- 1)))))
 (let (($x79 (or (or true $x71) $x78)))
 (let (($x95 (and (not (or false false)) $x79)))
 (and $x95 (= ?x81 0))))))))
(check-sat)

; 
(set-info :status unknown)
(declare-fun standard_metadata.ingress_port () (_ BitVec 9))
(assert
 (let (($x46 (= standard_metadata.ingress_port (_ bv1 9))))
 (or (or false (= standard_metadata.ingress_port (_ bv0 9))) $x46)))
(assert
 (let (($x71 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv1 1))))))
 (let (($x78 (and true (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv0 1))))))
 (let ((?x81 (ite $x78 0 (ite $x71 1 (- 1)))))
 (and (and (not (or false false)) (or (or true $x71) $x78)) (= ?x81 1))))))
(check-sat)

