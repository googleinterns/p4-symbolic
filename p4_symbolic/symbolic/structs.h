// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains structs used by our symbolic analysis.

#ifndef P4_SYMBOLIC_SYMBOLIC_STRUCTS_H_
#define P4_SYMBOLIC_SYMBOLIC_STRUCTS_H_

#include <memory>
#include <string>
#include <unordered_map>
#include <vector>

#include "google/protobuf/repeated_field.h"
#include "p4_pdpi/utils/status_utils.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {

// The result of analyzing an action and its body.
// It encapsulates a symbolic expression expressing its body's behavior in
// terms of its symbolic parameters.
// It provides a way to instantiate these parameters (and thus the body) for
// given concrete values.
class ActionSymbolicTrace {
 protected:
  // The fully qualified name of the corresponding action.
  std::string action_name_;
  // Z3 symbolic expressions (essentially symbolic variables) corresponding
  // to each parameter the action takes in order of declaration.
  std::vector<z3::expr> parameters_;
  // Symbolic expression expressing the semantic meaning of the action body,
  // over the symbolic parameters.
  z3::expr body_;

 public:
  ActionSymbolicTrace(std::string action_name, std::vector<z3::expr> parameters,
                      z3::expr body)
      : action_name_(action_name), parameters_(parameters), body_(body) {}

  // Instantiate the symbolic body defined over symbolic parameters with the
  // given values for these parameters.
  pdpi::StatusOr<z3::expr> instantiate(
      const google::protobuf::RepeatedField<int> &values) const {
    if (((unsigned int)values.size()) != this->parameters_.size()) {
      return absl::Status(
          absl::StatusCode::kInvalidArgument,
          absl::StrCat("Match entry has inconsistent number of parameters"));
    }

    z3::expr instantiated_call = this->body_;
    for (size_t i = 0; i < this->parameters_.size(); i++) {
      instantiated_call =
          instantiated_call && (this->parameters_.at(i) == values.at(i));
    }

    return instantiated_call;
  }
};

// The overall state of our symbolic analysis.
// This is returned by our main analysis function, and is used
// to find concrete test packets and for debugging.
struct SolverState {
  // The IR represnetation of the p4 program being analyzed.
  ir::P4Program program;
  // Z3 context wrapped in a smart pointer.
  // z3::context implements a private copy constructor and no move contructor.
  // This makes the call tostd::move(<context>) in pdpi::StatusOr
  // fail with a compile time error.
  // Z3's context is not meant to be copied or moved, related symbolic
  // expressions must be created using the exact same context.
  // Using a smart pointer address these issues and automatically manages the
  // memory.
  std::unique_ptr<z3::context> context;
  // Maps fully qualified action name to its symbolic trace, which can be used
  // to instatiate it for concrete parameter values.
  std::unordered_map<std::string, ActionSymbolicTrace> action_map;
  // Similar issue to context make unique_ptr desirable here.
  // Defining the solver here and re-using it for all future analysis allows
  // Z3 to remember interesting judgments it made about our trace constraints,
  // and also avoid storing and handling z3 symbolic constraints directly.
  std::unique_ptr<z3::solver> solver;
};

}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_STRUCTS_H_
