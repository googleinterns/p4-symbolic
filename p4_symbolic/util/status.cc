// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "p4_symbolic/util/status.h"

namespace p4_symbolic {
namespace util {

namespace pb_err = google::protobuf::util::error;

absl::Status ProtobufToAbslStatus(
    const google::protobuf::util::Status& status) {
  absl::StatusCode out_code;

  google::protobuf::util::error::Code code = status.code();
  switch (code) {
    case pb_err::OK:
      out_code = absl::StatusCode::kOk;
      break;
    case pb_err::CANCELLED:
      out_code = absl::StatusCode::kCancelled;
      break;
    case pb_err::UNKNOWN:
      out_code = absl::StatusCode::kUnknown;
      break;
    case pb_err::INVALID_ARGUMENT:
      out_code = absl::StatusCode::kInvalidArgument;
      break;
    case pb_err::DEADLINE_EXCEEDED:
      out_code = absl::StatusCode::kDeadlineExceeded;
      break;
    case pb_err::NOT_FOUND:
      out_code = absl::StatusCode::kNotFound;
      break;
    case pb_err::ALREADY_EXISTS:
      out_code = absl::StatusCode::kAlreadyExists;
      break;
    case pb_err::PERMISSION_DENIED:
      out_code = absl::StatusCode::kPermissionDenied;
      break;
    case pb_err::UNAUTHENTICATED:
      out_code = absl::StatusCode::kUnauthenticated;
      break;
    case pb_err::RESOURCE_EXHAUSTED:
      out_code = absl::StatusCode::kResourceExhausted;
      break;
    case pb_err::FAILED_PRECONDITION:
      out_code = absl::StatusCode::kFailedPrecondition;
      break;
    case pb_err::ABORTED:
      out_code = absl::StatusCode::kAborted;
      break;
    case pb_err::OUT_OF_RANGE:
      out_code = absl::StatusCode::kOutOfRange;
      break;
    case pb_err::UNIMPLEMENTED:
      out_code = absl::StatusCode::kUnimplemented;
      break;
    case pb_err::INTERNAL:
      out_code = absl::StatusCode::kInternal;
      break;
    case pb_err::UNAVAILABLE:
      out_code = absl::StatusCode::kUnavailable;
      break;
    case pb_err::DATA_LOSS:
      out_code = absl::StatusCode::kDataLoss;
      break;
    default:
      out_code = absl::StatusCode::kUnknown;
  }

  return absl::Status(out_code, status.error_message().ToString());
}

}  // namespace util
}  // namespace p4_symbolic
