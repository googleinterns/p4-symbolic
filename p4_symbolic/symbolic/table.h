// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically analyze P4 tables and their entries.
// A table is turned into a sequence of if-conditions (one per entry),
// each condition corresponds to having that entry matched on, and the
// corresponding then body invokes the appropriate symbolic action predicate
// with the parameters specified in the entry.

#ifndef P4_SYMBOLIC_SYMBOLIC_TABLE_H_
#define P4_SYMBOLIC_SYMBOLIC_TABLE_H_

#include "p4_pdpi/utils/status_utils.h"
#include "p4_symbolic/ir/ir.pb.h"
#include "p4_symbolic/symbolic/structs.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace table {

pdpi::StatusOr<z3::expr> AnalyzeTable(const ir::Table &table,
                                      const SolverState &solver_state);

}  // namespace table
}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_TABLE_H_
