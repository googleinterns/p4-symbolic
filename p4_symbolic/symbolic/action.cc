// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "p4_symbolic/symbolic/action.h"

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"

namespace p4_symbolic {
namespace symbolic {
namespace action {

gutil::StatusOr<SymbolicPerPacketState> EvaluateStatement(
    const ir::Statement &statement, const SymbolicPerPacketState &state,
    ActionContext *context) {
  switch (statement.statement_case()) {
    case ir::Statement::kAssignment:
      return EvaluateAssignmentStatement(statement.assignment(), state,
                                         context);

    default:
      return absl::Status(absl::StatusCode::kUnimplemented,
                          absl::StrCat("Action ", context->action_name,
                                       " contains unsupported statement ",
                                       statement.DebugString()));
  }
}

// Constructs a symbolic expression for the assignment value, and either
// constrains it in an enclosing assignment expression, or stores it in
// the action scope.
gutil::StatusOr<SymbolicPerPacketState> EvaluateAssignmentStatement(
    const ir::AssignmentStatement &assignment,
    const SymbolicPerPacketState &state, ActionContext *context) {
  // Evaluate RValue recursively, evaluate LValue in this function, then
  // assign RValue to the scope at LValue.
  ASSIGN_OR_RETURN(z3::expr right,
                   EvaluateRValue(assignment.right(), state, context));

  switch (assignment.left().lvalue_case()) {
    case ir::LValue::kFieldValue: {
      const ir::FieldValue &field_value = assignment.left().field_value();
      // TODO(babman): Better handling of metadata and headers in general.
      //               What should be hardcoded? probably only the headers
      //               used for packet generation.
      //               Need a better way to translate between hardcoded names
      //               and strings / FieldValue objects.
      //               Need to detect fields automatically. We already have this
      //               information in bmv2, and we already did it in a previous
      //               version of this code.
      SymbolicPerPacketState modified_state = state;
      std::string field_name = absl::StrFormat(
          "%s.%s", field_value.header_name(), field_value.field_name());
      if (modified_state.metadata.count(field_name) != 1) {
        return absl::Status(absl::StatusCode::kUnimplemented,
                            absl::StrCat("Action ", context->action_name,
                                         " refers to unknown header field ",
                                         field_value.DebugString()));
      }

      modified_state.metadata.at(field_name) = right;
      return modified_state;
    }

    case ir::LValue::kVariableValue: {
      const std::string &variable = assignment.left().variable_value().name();
      context->scope.insert_or_assign(variable, right);
      return state;
    }

    default:
      return absl::Status(absl::StatusCode::kUnimplemented,
                          absl::StrCat("Action ", context->action_name,
                                       " contains unsupported LValue ",
                                       assignment.left().DebugString()));
  }
}

// Constructs a symbolic expression corresponding to this value, according
// to its type.
gutil::StatusOr<z3::expr> EvaluateRValue(const ir::RValue &rvalue,
                                         const SymbolicPerPacketState &state,
                                         ActionContext *context) {
  switch (rvalue.rvalue_case()) {
    case ir::RValue::kFieldValue:
      return EvaluateFieldValue(rvalue.field_value(), state, context);

    case ir::RValue::kVariableValue:
      return EvaluateVariable(rvalue.variable_value(), state, context);

    default:
      return absl::Status(
          absl::StatusCode::kUnimplemented,
          absl::StrCat("Action ", context->action_name,
                       " contains unsupported RValue ", rvalue.DebugString()));
  }
}

// Extract the field symbolic value from the symbolic state.
gutil::StatusOr<z3::expr> EvaluateFieldValue(
    const ir::FieldValue &field_value, const SymbolicPerPacketState &state,
    ActionContext *context) {
  std::string field_name = absl::StrFormat("%s.%s", field_value.header_name(),
                                           field_value.field_name());
  if (state.metadata.count(field_name) != 1) {
    return absl::Status(absl::StatusCode::kUnimplemented,
                        absl::StrCat("Action ", context->action_name,
                                     " refers to unknown header field ",
                                     field_value.DebugString()));
  }

  return state.metadata.at(field_name);
}

// Looks up the symbolic value of the variable in the action scope.
gutil::StatusOr<z3::expr> EvaluateVariable(const ir::Variable &variable,
                                           const SymbolicPerPacketState &state,
                                           ActionContext *context) {
  std::string variable_name = variable.name();
  if (context->scope.count(variable_name) != 1) {
    return absl::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("Action ", context->action_name,
                     " refers to undefined variable ", variable_name));
  }

  return context->scope.at(variable_name);
}

// Symbolically evaluates the given action on the given symbolic parameters.
// This produces a symbolic expression on the symbolic parameters that is
// semantically equivalent to the behavior of the action on its concrete
// parameters.
gutil::StatusOr<SymbolicPerPacketState> EvaluateAction(
    const ir::Action &action, const google::protobuf::RepeatedField<int> &args,
    const SymbolicPerPacketState &state) {
  // Construct this action's context.
  ActionContext context;
  context.action_name = action.action_definition().preamble().name();

  // Add action parameters to scope.
  const auto &parameters = action.action_definition().params_by_id();
  if (static_cast<int>(parameters.size()) != args.size()) {
    return absl::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("Action ", action.action_definition().preamble().name(),
                     " called with incompatible number of parameters"));
  }

  for (size_t i = 1; i <= parameters.size(); i++) {  // In order of definition.
    const pdpi::IrActionDefinition::IrActionParamDefinition &parameter =
        parameters.at(i);
    const std::string &parameter_name = parameter.param().name();
    z3::expr parameter_value = Z3Context().int_val(args.at(i - 1));
    context.scope.insert({parameter_name, parameter_value});
  }

  // Iterate over the body in order, and evaluate each statement.
  SymbolicPerPacketState result = state;
  for (const auto &statement : action.action_implementation().action_body()) {
    ASSIGN_OR_RETURN(result, EvaluateStatement(statement, result, &context));
  }

  return result;
}

}  // namespace action
}  // namespace symbolic
}  // namespace p4_symbolic
