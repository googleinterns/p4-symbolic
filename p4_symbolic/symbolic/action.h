// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically analyze/interprate actions
// and their bodies.
// An action is represented as a boolean symbolic z3 expression over
// unconstrained symbolic parameters corresponding to its actual P4 parameters.

#ifndef P4_SYMBOLIC_SYMBOLIC_ACTION_H_
#define P4_SYMBOLIC_SYMBOLIC_ACTION_H_

#include <string>
#include <unordered_map>
#include <vector>

#include "p4_pdpi/utils/status_utils.h"
#include "p4_symbolic/ir/ir.pb.h"
#include "p4_symbolic/symbolic/symbolic.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace action {

// Symbolically runs the given action on the given symbolic parameters.
// This produces a symbolic expression on the symbolic parameters that is
// semantically equivalent to the behavior of the action on its concrete
// parameters.
pdpi::StatusOr<SymbolicState> RunAction(
    const ir::Action &action, const std::vector<z3::expr> &symbolic_parameters,
    const SymbolicState &state);

// Internal functions used to run statements and expressions within an action
// body. These are internal functions not used beyond this header and its
// associated source file.

// The scope of this action: maps variable names to their symbolic values.
using ActionContext = std::unordered_map<std::string, z3::expr>;

// Performs a switch case over support statement types and call the
// appropriate function.
pdpi::StatusOr<SymbolicState> RunStatement(const ir::Statement &statement,
                                           const SymbolicState &state,
                                           ActionContext *context);

// Constructs a symbolic expression for the assignment value, and either
// constrains it in an enclosing assignment expression, or stores it in
// the action scope.
pdpi::StatusOr<SymbolicState> RunAssignmentStatement(
    const ir::AssignmentStatement &assignment, const SymbolicState &state,
    ActionContext *context);

// Constructs a symbolic expression corresponding to this value, according
// to its type.
pdpi::StatusOr<z3::expr> RunRValue(const ir::RValue &rvalue,
                                   const SymbolicState &state,
                                   ActionContext *context);

// Extract the field symbolic value from the symbolic state.
pdpi::StatusOr<z3::expr> RunFieldValue(const ir::FieldValue &field_value,
                                       const SymbolicState &state,
                                       ActionContext *context);

// Looks up the symbolic value of the variable in the action scope.
pdpi::StatusOr<z3::expr> RunVariable(const ir::Variable &variable,
                                     const SymbolicState &state,
                                     ActionContext *context);

}  // namespace action
}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_ACTION_H_
