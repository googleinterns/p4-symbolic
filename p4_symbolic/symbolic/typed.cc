// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "p4_symbolic/symbolic/typed.h"

#include <utility>

#include "p4_symbolic/symbolic/symbolic.h"

// Pad if needed by using the appropriate z3 API.
#define PAD(bitvector, pad_size)                                               \
  pad_size > 0 ? z3::concat(Z3Context().bv_val(0, pad_size), bitvector.expr()) \
               : bitvector.expr()

// This macro is used through out the implementation of the overloaded
// operators for the TypedExpr class.
// It sort checks the two operands a and b, and if they were bit vectors,
// it will perform any needed size padding.
// The macro will define two z3::expr variables "a_expr" and "b_expr",
// corresponding to the finalized expressions of the two operands.
#define SORT_CHECK_AND_PAD(a, b)                        \
  z3::expr a_expr = a.expr_;                            \
  z3::expr b_expr = b.expr_;                            \
  assert(a.sort().sort_kind() == b.sort().sort_kind()); \
  if (a.sort().is_bv()) {                               \
    int a_len = a.sort().bv_size();                     \
    int b_len = b.sort().bv_size();                     \
    a_expr = PAD(a, b_len - a_len);                     \
    b_expr = PAD(b, a_len - b_len);                     \
  }

namespace p4_symbolic {
namespace symbolic {

// Copy and move assignment operators must respect pre-declared sort.
TypedExpr &TypedExpr::operator=(const TypedExpr &other) {
  assert(this->sort().sort_kind() == other.sort().sort_kind());
  if (this->sort().is_bv()) {
    // We can only re-assign bit vectors to ones of equal or larger size.
    unsigned int size_diff = this->sort().bv_size() - other.sort().bv_size();
    assert(size_diff >= 0);
    // All is good, copy other to this padding it to the proper bit length.
    this->expr_ = PAD(other, size_diff);
  } else {
    // For other sorts, we are all good and can assign directly.
    this->expr_ = other.expr_;
  }

  return *this;
}
TypedExpr &TypedExpr::operator=(TypedExpr &&other) {
  assert(this->sort().sort_kind() == other.sort().sort_kind());
  if (this->sort().is_bv()) {
    unsigned int size_diff = this->sort().bv_size() - other.sort().bv_size();
    assert(size_diff >= 0);
    this->expr_ = std::move(PAD(other, size_diff));
  } else {
    this->expr_ = std::move(other.expr_);
  }
  return *this;
}

// Overloaded operators.
// Arithmetic operators.
TypedExpr TypedExpr::operator+(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr + b_expr);
}
TypedExpr TypedExpr::operator-(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr - b_expr);
}
TypedExpr TypedExpr::operator*(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr * b_expr);
}

// Relational operators.
TypedExpr TypedExpr::operator==(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr == b_expr);
}
TypedExpr TypedExpr::operator!=(const TypedExpr &b) const {
  return !(*this == b);
}
TypedExpr TypedExpr::operator<(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(z3::ult(a_expr, b_expr));
}
TypedExpr TypedExpr::operator<=(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(z3::ule(a_expr, b_expr));
}
TypedExpr TypedExpr::operator>(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(z3::ugt(a_expr, b_expr));
}
TypedExpr TypedExpr::operator>=(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(z3::uge(a_expr, b_expr));
}

// Boolean operators.
TypedExpr TypedExpr::operator!() const { return TypedExpr(!this->expr_); }
TypedExpr TypedExpr::operator&&(const TypedExpr &b) const {
  return TypedExpr(this->expr_ && b.expr_);
}
TypedExpr TypedExpr::operator||(const TypedExpr &b) const {
  return TypedExpr(this->expr_ || b.expr_);
}

// Binary operators.
TypedExpr TypedExpr::operator~() const { return TypedExpr(~this->expr_); }
TypedExpr TypedExpr::operator&(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr & b_expr);
}
TypedExpr TypedExpr::operator|(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr | b_expr);
}
TypedExpr TypedExpr::operator^(const TypedExpr &b) const {
  SORT_CHECK_AND_PAD((*this), b);
  return TypedExpr(a_expr ^ b_expr);
}

// If-then-else.
TypedExpr TypedExpr::ite(const TypedExpr &condition,
                         const TypedExpr &true_value,
                         const TypedExpr &false_value) {
  // Optimization: if both branches are the same *syntactically*
  // then we can just use the branch expression directly and no need
  // for an if-then-else statement.
  if (z3::eq(true_value.expr(), false_value.expr())) {
    return true_value;
  }

  // Values in both cases must have the same sort and signedness.
  SORT_CHECK_AND_PAD(true_value, false_value);
  return TypedExpr(z3::ite(condition.expr_, a_expr, b_expr));
}

// Bit-wise shifts
TypedExpr TypedExpr::shl(const TypedExpr &bit_vector,
                         const TypedExpr &shift_value) {
  return TypedExpr(z3::shl(bit_vector.expr_, shift_value.expr_));
}
TypedExpr TypedExpr::shr(const TypedExpr &bit_vector,
                         const TypedExpr &shift_value) {
  return TypedExpr(z3::lshr(bit_vector.expr_, shift_value.expr_));
}

}  // namespace symbolic
}  // namespace p4_symbolic
