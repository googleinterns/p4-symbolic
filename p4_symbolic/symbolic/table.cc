// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically evaluate P4 tables and their entries.
// A table is turned into a sequence of if-conditions (one per entry),
// each condition corresponds to having that entry matched on, and the
// corresponding then body invokes the appropriate symbolic action expression
// with the parameters specified in the entry.

#include "p4_symbolic/symbolic/table.h"

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "p4_symbolic/symbolic/action.h"
#include "p4_symbolic/symbolic/typed.h"
#include "p4_symbolic/symbolic/util.h"

namespace p4_symbolic {
namespace symbolic {
namespace table {

namespace {

// Analyze a single match that is part of a table entry.
// Constructs a symbolic expression that semantically corresponds to this
// match.
gutil::StatusOr<TypedExpr> AnalyzeSingleMatch(
    p4::config::v1::MatchField match_definition,
    const TypedExpr &field_expression, const pdpi::IrMatch &match) {
  if (match_definition.match_case() != p4::config::v1::MatchField::kMatchType) {
    // Arch-specific match type.
    return absl::InvalidArgumentError(
        absl::StrCat("Found match with non-standard type"));
  }

  // TODO(babman): Support the other match types.
  switch (match_definition.match_type()) {
    case p4::config::v1::MatchField::EXACT: {
      if (match.match_value_case() != pdpi::IrMatch::kExact) {
        return absl::InvalidArgumentError(
            absl::StrCat("Match definition in table has type \"Exact\" but its "
                         "invocation in TableEntry has a different type ",
                         match_definition.DebugString()));
      }
      ASSIGN_OR_RETURN(TypedExpr value_expression,
                       util::IrValueToZ3Expr(match.exact()));
      return field_expression == value_expression;
    }
    default:
      return absl::UnimplementedError(absl::StrCat(
          "Found unsupported match type ", match_definition.DebugString()));
  }
}

// Constructs a symbolic expression that is true if and only if this entry
// is matched on.
gutil::StatusOr<TypedExpr> AnalyzeTableEntryCondition(
    const ir::Table &table, const pdpi::IrTableEntry &entry,
    const SymbolicHeaders &headers) {
  // Make sure number of match keys is the same in the table definition and
  // table entry.
  if (table.table_definition().match_fields_by_id_size() !=
      entry.matches_size()) {
    return absl::InvalidArgumentError(
        absl::StrCat("Found a match entry ", entry.DebugString(), " in table",
                     table.table_definition().preamble().name(),
                     " with wrong match fields count"));
  }

  // Construct the match condition expression.
  TypedExpr condition_expression = TypedExpr(Z3Context().bool_val(true));
  for (const auto &[id, match_fields] :
       table.table_definition().match_fields_by_id()) {
    p4::config::v1::MatchField match_definition = match_fields.match_field();

    // Match id is unique only within the enclosing table.
    // https://github.com/p4lang/p4runtime/blob/master/docs/v1/p4runtime-id-notes.md
    // Match id starts at 1, and proceeds in the same order the matches are
    // defined in, which is identical to the order they are provided by in
    // the table entries file.
    const pdpi::IrMatch &match = entry.matches(id - 1);

    // TODO(babman): We should put in the expression of the match from bmv2
    //               json in the IR, and use instead of the name.
    //               @konne mentioned this before in a meeting.
    const std::string &match_string_expression = match_definition.name();
    if (headers.count(match_string_expression) != 1) {
      return absl::InvalidArgumentError(absl::StrCat(
          "Unknown table match expression ", match_string_expression,
          " in table ", table.table_definition().preamble().name()));
    }
    ASSIGN_OR_RETURN(
        TypedExpr match_expression,
        AnalyzeSingleMatch(match_definition,
                           headers.at(match_string_expression), match));
    condition_expression = condition_expression && match_expression;
  }

  return condition_expression;
}

// Constructs a symbolic expressions that represents the action invocation
// corresponding to this entry.
gutil::StatusOr<SymbolicHeaders> AnalyzeTableEntryAction(
    const ir::Table &table, const pdpi::IrTableEntry &entry,
    const google::protobuf::Map<std::string, ir::Action> &actions,
    const SymbolicHeaders &headers) {
  // Check that the action invoked by entry exists.
  const std::string &table_name = table.table_definition().preamble().name();
  const std::string &action_name = entry.action().name();
  if (actions.count(action_name) != 1) {
    return absl::InvalidArgumentError(
        absl::StrCat("Found a match entry ", entry.DebugString(), " in table",
                     table_name, " referring to unknown action ", action_name));
  }

  // Instantiate the action's symbolic expression with the entry values.
  const ir::Action &action = actions.at(action_name);
  return action::EvaluateAction(action, entry.action().params(), headers);
}

}  // namespace

gutil::StatusOr<SymbolicHeadersAndTableMatch> EvaluateTable(
    const ir::Table &table, const std::vector<pdpi::IrTableEntry> &entries,
    const google::protobuf::Map<std::string, ir::Action> &actions,
    const SymbolicHeaders &headers) {
  // The overall structure describing the match on this table.
  SymbolicTableMatch table_match = {
      TypedExpr(Z3Context().bool_val(false)),  // No match yet!
      TypedExpr(Z3Context().int_val(-1)),      // No match index.
      TypedExpr(Z3Context().int_val(-1))       // TODO(babman): bitvector.
  };

  // Begin with the default entry.
  pdpi::IrTableEntry default_entry;
  default_entry.mutable_action()->set_name(
      table.table_implementation().default_action());
  for (const std::string &parameter_value :
       table.table_implementation().default_action_parameters()) {
    ASSIGN_OR_RETURN(
        *(default_entry.mutable_action()->add_params()->mutable_value()),
        util::StringToIrValue(parameter_value));
  }
  ASSIGN_OR_RETURN(
      SymbolicHeaders modified_headers,
      AnalyzeTableEntryAction(table, default_entry, actions, headers));

  // The table semantically is just a bunch of if conditions, one per
  // table entry, we construct this big if-elseif-...-else construct via
  // this simpler representation.
  // Traverse in reverse order: from least to highest priority, since we are
  // building the if-elseif-...-else statement inside out.
  for (int row = entries.size() - 1; row >= 0; row--) {
    const pdpi::IrTableEntry &entry = entries.at(row);
    ASSIGN_OR_RETURN(TypedExpr row_match,
                     AnalyzeTableEntryCondition(table, entry, headers));
    ASSIGN_OR_RETURN(SymbolicHeaders row_headers,
                     AnalyzeTableEntryAction(table, entry, actions, headers));

    // Using this alias makes it simpler to put constraints on packet later.
    table_match.matched = table_match.matched || row_match;
    table_match.entry_index =
        TypedExpr::ite(row_match, TypedExpr(Z3Context().int_val(row)),
                       table_match.entry_index);

    // Apply the changes to the header fields if the symbolic condition to match
    // on this entry is satisified.
    modified_headers =
        util::MergeHeadersOnCondition(modified_headers, row_headers, row_match);
  }

  return SymbolicHeadersAndTableMatch{modified_headers, table_match};
}

}  // namespace table
}  // namespace symbolic
}  // namespace p4_symbolic
