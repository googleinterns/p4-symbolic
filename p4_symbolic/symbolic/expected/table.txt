Finding packet for table MyIngress.ports_exact and row 0
	standard_metadata.ingress_port = 0
	standard_metadata.egress_spec = 1

Finding packet for table MyIngress.ports_exact and row 1
	standard_metadata.ingress_port = 1
	standard_metadata.egress_spec = 0

